README.md
# thunk-redis

The fastest thunk/promise-based redis client, support all redis features.

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Downloads][downloads-image]][downloads-url]
[![js-standard-style][js-standard-image]][js-standard-url]

## [thunks](https://github.com/thunks/thunks)

## Implementations:

- [thunk-ratelimiter](https://github.com/thunks/thunk-ratelimiter) The fastest abstract rate limiter.
- [timed-queue](https://github.com/teambition/timed-queue) Distributed timed job queue, backed by Redis.

## Demo([examples](https://github.com/zensh/thunk-redis/blob/master/examples))

**https://raw.githubusercontent.com/antirez/redis/4.0/redis.conf**

**Sugest set config `cluster-require-full-coverage` to `no` in redis cluster!**

**default thunk API:**

```js
const redis = require('../index')
const thunk = require('thunks')()
const client = redis.createClient({
  database: 1
})

client.on('connect', function () {
  console.log('redis connected!')
})

client.info('server')(function (error, res) {
  console.log('redis server info:', res)
  return this.dbsize()
})(function (error, res) {
  console.log('current database size:', res)
  // current database size: 0
  return this.select(0)
})(function (error, res) {
  console.log('select database 0:', res)
  // select database 0: OK
  return thunk.all([
    this.multi(),
    this.set('key', 'redis'),
    this.get('key'),
    this.exec()
  ])
})(function (error, res) {
  console.log('transactions:', res)
  // transactions: [ 'OK', 'QUEUED', 'QUEUED', [ 'OK', 'redis' ] ]
  return this.quit()
})(function (error, res) {
  console.log('redis client quit:', res)
  // redis client quit: OK
})
```

**use promise API:**
```js
const redis = require('../index')
const Promise = require('bluebird')
const client = redis.createClient({
  database: 1,
  usePromise: true
})

client.on('connect', function () {
  console.log('redis connected!')
})

client
  .info('server')
  .then(function (res) {
    console.log('redis server info:', res)
    return client.dbsize()
  })
  .then(function (res) {
    console.log('current database size:', res)
    // current database size: 0
    return client.select(0)
  })
  .then(function (res) {
    console.log('select database 0:', res)
    // select database 0: OK
    return Promise.all([
      client.multi(),
      client.set('key', 'redis'),
      client.get('key'),
      client.exec()
    ])
  })
  .then(function (res) {
    console.log('transactions:', res)
    // transactions: [ 'OK', 'QUEUED', 'QUEUED', [ 'OK', 'redis' ] ]
    return client.quit()
  })
  .then(function (res) {
    console.log('redis client quit:', res)
    // redis client quit: OK
  })
  .catch(function (err) {
    console.error(err)
  })
```

**support generator in thunk API:**
```js
const redis = require('thunk-redis')
const client = redis.createClient()

client.select(1)(function* (error, res) {
  console.log(error, res)

  yield this.set('foo', 'bar')
  yield this.set('bar', 'baz')

  console.log('foo -> %s', yield this.get('foo'))
  console.log('bar -> %s', yield this.get('bar'))

  var user = {
    id: 'u001',
    name: 'jay',
    age: 24
  }
  // transaction, it is different from node_redis!
  yield [
    this.multi(),
    this.set(user.id, JSON.stringify(user)),
    this.zadd('userAge', user.age, user.id),
    this.pfadd('ageLog', user.age),
    this.exec()
  ]

  return this.quit()
})(function (error, res) {
  console.log(error, res)
})
```

## Benchmark

Details: https://github.com/thunks/thunk-redis/issues/12

## Installation

**Node.js:**

```bash
npm install thunk-redis
```

## API ([More](https://github.com/zensh/thunk-redis/blob/master/API.md))

1. redis.createClient([addressArray], [options])
2. redis.createClient([port], [host], [options])
3. redis.createClient([address], [options])
4. redis.calcSlot(str)
5. redis.log([...])

### redis.log

Helper tool, print result or error stack.

```js
const client = redis.createClient()
client.info()(redis.log)
```

### redis.createClient

```js
const client1 = redis.createClient()
const client2 = redis.createClient({database: 2})
const client3 = redis.createClient(6379, {database: 2})
const client4 = redis.createClient('127.0.0.1:6379', {database: 2})
const client5 = redis.createClient(6379, '127.0.0.1', {database: 2})
// connect to 2 nodes
const client6 = redis.createClient([6379, 6380])
const client7 = redis.createClient(['127.0.0.1:6379', '127.0.0.1:6380']) // IPv4
const client8 = redis.createClient(['[::1]:6379', '[::1]:6380']) // IPv6
```

**redis cluster:**

```js
// assume cluster: '127.0.0.1:7000', '127.0.0.1:7001', '127.0.0.1:7002', ...
const client1 = redis.createClient(7000, options) // will auto find cluster nodes!
const client2 = redis.createClient([7000, 7001, 7002], options)

const client3 = redis.createClient([
  '127.0.0.1:7000',
  '127.0.0.1:7001',
  '127.0.0.1:7002'
], options)

const client4 = redis.createClient([
  {host: '127.0.0.1', port: 7000},
  {host: '127.0.0.1', port: 7001},
  {host: '127.0.0.1', port: 7002},
], options)
// All of above will work, it will find redis nodes by self.

// Create a client in cluster servers without cluster mode:
const clientX = redis.createClient(7000, {clusterMode: false})
```

- `options.onlyMaster`: *Optional*, Type: `Boolean`, Default: `true`.

    In replication mode, thunk-redis will try to connect master node and close slave node.

- `options.authPass`: *Optional*, Type: `String`, Default: `''`.

- `options.database`: *Optional*, Type: `Number`, Default: `0`.

- `options.returnBuffers`: *Optional*, Type: `Boolean`, Default: `false`.

- `options.usePromise`: *Optional*, Type: `Boolean`, Default: `false`.

    Export promise commands API.

    **Use default Promise:**
    ```js
    var redis = require('thunk-redis')
    var client = redis.createClient({
      usePromise: true
    })
    ```

- `options.noDelay`: *Optional*, Type: `Boolean`, Default: `true`.

    Disables the Nagle algorithm. By default TCP connections use the Nagle algorithm, they buffer data before sending it off. Setting true for noDelay will immediately fire off data each time socket.write() is called.

- `options.retryMaxDelay`: *Optional*, Type: `Number`, Default: `5 * 60 * 1000`.

    By default every time the client tries to connect and fails time before reconnection (delay) almost multiply by `1.2`. This delay normally grows infinitely, but setting `retryMaxDelay` limits delay to maximum value, provided in milliseconds.

- `options.maxAttempts`: *Optional*, Type: `Number`, Default: `20`.

    By default client will try reconnecting until connected. Setting `maxAttempts` limits total amount of reconnects.

- `options.pingInterval`: *Optional*, Type: `Number`, Default: `0`.

    How many ms before sending a ping packet. There is no ping packet by default(`0` to disable). If redis server enable `timeout` config, this option will be useful.

- `options.IPMap`: *Optional*, Type: `Object`, Default: `{}`.

    This option use o resolve redis internal IP and external IP problem https://github.com/thunks/thunk-redis/issues/19. Define it like: `{internalIP: externalIP}`, for example:
    ```js
    const cli = redis.createClient([
      '127.0.0.1:7000',
      '127.0.0.1:7001',
      '127.0.0.1:7002',
      '127.0.0.1:7003',
      '127.0.0.1:7004',
      '127.0.0.1:7005'
    ], {
      IPMap: {
        '172.17.0.2:7000': '127.0.0.1:7000',
        '172.17.0.2:7001': '127.0.0.1:7001',
        '172.17.0.2:7002': '127.0.0.1:7002',
        '172.17.0.2:7003': '127.0.0.1:7003',
        '172.17.0.2:7004': '127.0.0.1:7004',
        '172.17.0.2:7005': '127.0.0.1:7005'
      }
    })
    ```

[npm-url]: https://npmjs.org/package/thunk-redis
[npm-image]: http://img.shields.io/npm/v/thunk-redis.svg

[travis-url]: https://travis-ci.org/thunks/thunk-redis
[travis-image]: http://img.shields.io/travis/thunks/thunk-redis.svg

[downloads-url]: https://npmjs.org/package/thunk-redis
[downloads-image]: http://img.shields.io/npm/dm/thunk-redis.svg?style=flat-square

[js-standard-url]: https://github.com/feross/standard
[js-standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file starting from version **v1.1.0**.
This project adheres to [Semantic Versioning](http://semver.org/).

-----
## [2.1.7, 2.2.1] - 2018-09-27

### Changed

- support redisUrl array.

## [2.2.0] - 2018-07-01

### Changed

- support redis v5 commands.

## [2.1.6] - 2017-10-12

### Fixed

- fix evalauto command.

## [2.1.4] - 2017-10-12

### Fixed

- Update dependencies.
- fix evalauto command.

## [2.1.1] - 2017-06-03

### Changed

- Update dependencies.
- Add Node.js v8 to test.
- Update dependencies, changed node engines to >= v4.5.0.

## [2.1.0] - 2017-04-27

### Changed

- Improve code with ES2015 style.
- remove `each` function and `slice` function.

## [2.0.0] - 2017-03-30

### Changed

- Update dependencies, changed node engines to >= v4.

## [1.7.3] - 2016-10-27

### Changed

- Update dependencies.

### Fixed

- Fixed for docker-redis-cluster, https://github.com/thunks/thunk-redis/issues/19.

## [1.7.2] - 2016-10-02

### Changed

- Update dependencies.

## [1.7.1] - 2016-09-14

### Changed

- Update dependencies.

## [1.7.0] - 2016-08-21

### Changed

- Update engines to ">=0.12".

## [1.6.7] - 2016-08-15

### Changed

- Update dependencies.
- Improved socket send.

## [1.6.6] - 2016-07-22

### Changed

- Update dependencies.
- Added "TOUCH" command.

## [1.6.5] - 2016-06-08

### Changed

- #18 support redis URL: redis://USER:PASS@redis.com:5678

## [1.6.4] - 2016-06-07

### Changed

- Update dependencies.

## [1.6.3] - 2016-05-29

### Changed

- Improved performance.

## [1.6.2] - 2016-05-24

### Fixed

- Fixed for PubSub keyword 'message'.

## [1.6.1] - 2016-05-24

### Changed

- Removed "debug" module.
- Update dependencies.

## [1.6.0] - 2016-05-07

### Changed

- Supported redis v3.2.

## [1.5.4] - 2016-03-16

### Changed

- Removed socket.cork account to a bug in node.js.

## [1.5.3] - 2016-03-14

### Changed

- Update dependencies.

## [1.5.2] - 2016-03-01

### Changed

- Improve connection.

### Fixed

- Fixed for old redis(v2.8.x).

## [1.5.1] - 2016-02-21

### Changed

- Update dependencies.
- Improved code.

## [1.5.0] - 2016-02-18

### Changed

- Removed auto-discover cluster nodes during initialization. Because the nodes information
from "cluster slots" command includes local-host information. But it will anto-connect
node by "MOVED" and "ASK".
- Changed files structure.

## [1.4.1] - 2016-01-20

### Changed

- Update dependencies.

## [1.4.0] - 2015-12-29

### Changed

- Changed default `options.maxAttempts` to `5`.
- Added `options.onlyMaster`, it is useful for replication mode.
- Removed `options.handleError`.
- Supported IPv6.

## [1.3.0] - 2015-12-13

### Changed

- Added `options.pingInterval`.
- Added `client.clientConnect`.

## [1.2.4] - 2015-11-29

### Changed

- Update dependencies.

## [1.2.3] - 2015-11-27

### Fixed

- Fixed for slave node.

## [1.2.2] - 2015-11-19

### Changed

- Update description.

## [1.2.1] - 2015-11-18

### Changed

- Improved performance.

## [1.2.0] - 2015-11-12

### Changed

- Update `thunks` to v4.0.0.

## [1.1.1] - 2015-10-07

### Changed

- Added `clientReady` method.

### Fixed

- Fixed command `evalauto`.
- Fixed for `ASK`.

## [1.1.0] - 2015-10-07

### Changed

- Added custom command `evalauto`.
- Used `const` instead of `var`.


lib/socket.js
'use strict';

var net = require('net');
var util = require('util');
var resp = require('respjs');

var tool = require('./tool');
var Queue = require('./queue');

var thunks = require('thunks');
var Thunk = thunks();

exports.sendCommand = sendCommand;
exports.createSockets = createSockets;

function createSockets(redis, optionsArray) {
  tool.each(optionsArray, function(option) {
    createSocket(redis, option, {
      attempts: 1,
      commandQueue: new Queue()
    });
  });
}

function addSocket(redis, redisId) {
  redisId = redisId.split(':');
  return createSocket(redis, {
    host: redisId[0],
    port: +redisId[1]
  }, {
    attempts: 1,
    commandQueue: new Queue()
  });
}

function createSocket(redis, options, socketState) {
  var socket, redisState = redis._redisState;
  if (!options.port || !options.host) socket = net.createConnection(options);
  else {
    var redisId = options.host + ':' + options.port;
    if (redisState.pool[redisId]) return redisState.pool[redisId];
    socket = redisState.pool[redisId] = net.createConnection(options);
  }

  socket.setNoDelay(redisState.options.noDelay);
  socket.setTimeout(redisState.options.timeout);
  socket.setKeepAlive(redisState.options.keepAlive);

  socket.pendingWatcher = null;
  socket.debugMode = redisState.options.debugMode;
  socket.returnBuffers = redisState.options.returnBuffers;
  socket.commandsHighWater = redisState.options.commandsHighWater;

  tool.setPrivate(socket, 'commandQueue', socketState.commandQueue);

  return socket
    .on('connect', function() {
      socket.connected = true;
      socket.redisId = socket.remoteAddress + ':' + socket.remotePort;
      checkConnection(redis, socket);
    })
    .on('data', function(chunk) {
      var reply = socket.pendingWatcher;
      if (socket.debugMode) tool.log({
        socketChunk: chunk,
        address: socket.redisId
      });
      if (!reply) return redis.emit('error', new Error('Unexpected reply: ' + chunk));
      if (!reply.resp) reply.resp = createResp(redis, socket, reply);
      reply.resp.feed(chunk);
    })
    .on('error', function(error) {
      flushToRedis(redis, this, error);
      redis.emit('error', error);
    })
    .on('close', function(hadError) {
      flushToRedis(redis, this, new Error(this.redisId + ' was closed'));
      unmountSocket(redis, socket, hadError, true);
    })
    .on('timeout', function() {
      var error = new Error('The redis connection was timeout');
      redis.emit('timeout');
      if (socketState.attempts <= redisState.options.maxAttempts) {
        flushPending(socket, error);
        unmountSocket(redis, socket, error);

        redis.emit('reconnecting', {
          delay: redisState.options.retryDelay,
          attempts: ++socketState.attempts
        });

        setTimeout(function() {
          createSocket(redis, options, {
            attempts: socketState.attempts,
            commandQueue: socket.commandQueue,
          });
        }, redisState.options.retryDelay);
      } else {
        flushToRedis(redis, socket, error);
        unmountSocket(redis, socket, error, true);
      }
    })
    .on('end', function() {
      flushToRedis(redis, this, new Error(socket.redisId + ' was ended'), true);
      unmountSocket(redis, socket);
    });
}

function checkConnection(redis, socket) {
  var redisState = redis._redisState;
  var _socket = redisState.pool[socket.redisId];
  if (_socket && _socket !== socket) {
    socket.destroy();
    redis.emit('warn', new Error('Create connection repeatedly, drop it: ' + socket.redisId));
    return;
  }

  redisState.pool[socket.redisId] = socket;

  var Thunk = thunks(function(error) {
    redis.emit('error', error);
  });

  Thunk(function(callback) {
    // auth
    if (!redisState.options.authPass) return callback();
    var command = createCommand(redis, ['auth', redisState.options.authPass], function(error, res) {
      if (res && res.toString() === 'OK') return callback();
      callback(new Error('Auth failed: ' + socket.redisId));
    });
    socket.commandQueue.push(command);

  })(function() {
    // check replication and cluster
    return function(callback) {
      var command = createCommand(redis, ['info', 'default'], function(error, res) {
        if (!res) return callback(error);
        res = res.toString();
        redisState.clusterMode = res.indexOf('cluster_enabled:1') > 0;
        socket.redisMaster = res.indexOf('role:master') > 0;
        if (socket.redisMaster && !redisState.slots[-1]) redisState.slots[-1] = socket.redisId;
        callback();
      });
      socket.commandQueue.push(command);
    };

  })(function() {
    // check cluster slots and connect them.
    return redisState.clusterMode && function(callback) {
      var command = createCommand(redis, ['cluster', 'slots'], function(error, res) {
        if (error) return callback(error);
        var netOptions = [];
        tool.each(res, function(info) {
          // [ 5461, 10922, [ '127.0.0.1', 7001 ], [ '127.0.0.1', 7004 ] ]
          var masterId, i = 1, replicaIds = [];

          while (info[++i]) {
            netOptions.push({
              host: info[i][0],
              port: info[i][1]
            });
            replicaIds.push(info[i][0] + ':' + info[i][1]);
          }
          masterId = replicaIds[0];
          if (redisState.pool[masterId]) redisState.pool[masterId].replicaIds = replicaIds.slice(1);

          for (i = info[0]; i <= info[1]; i++) redisState.slots[i] = masterId;
        });
        createSockets(redis, netOptions);
        callback();
      });
      socket.commandQueue.push(command);
    };

  })(function() {
    // check selected database
    if (redisState.clusterMode || !redisState.options.database) return;
    return function(callback) {
      var command = createCommand(redis, ['select', redisState.options.database], function(error, res) {
        if (error) return callback(error);
        redisState.database = redisState.options.database;
        callback();
      });
      socket.commandQueue.push(command);
    };

  })(function() {
    tool.log.call(redisState.options, socket.redisId + ' connected.');
    if (!redisState.slots[-1]) return;
    // default socket connected
    if (!redisState.connected) {
      redisState.connected = true;
      redis.emit('connect');
      dispatchCommands(redis);
    }
    execQueue(redisState, socket);
  });

  execQueue(redisState, socket);
}

function createResp(redis, socket, reply) {
  var redisState = redis._redisState;

  return new resp.Resp({
      returnBuffers: socket.returnBuffers,
      expectResCount: reply.commands.length
    })
    .on('error', function(error) {
      if (socket.debugMode) tool.log({respError: error});
      flushPending(socket, error);
      redis.emit('error', error);
    })
    .on('data', function(data) {
      if (socket.debugMode) tool.log({respData: data});

      var command = reply.commands.first();
      if (redisState.monitorMode && (!command || command.name !== 'quit'))
        return redis.emit('monitor', data);

      if (isMessageReply(data)) return redis.emit.apply(redis, data);

      if (isUnSubReply(data)) {
        if (redisState.pubSubMode && !data[2]) {
          redisState.pubSubMode = false;
          if (command) this.setAutoEnd(reply.commands.length);
        }

        if (!command) this.end();
        else if (data[0] === command.name) {
          reply.commands.shift();
          command.callback();
        }

        return redis.emit.apply(redis, data);
      }

      reply.commands.shift();
      if (!command) return redis.emit('error', new Error('Unexpected reply: ' + data));

      if (util.isError(data)) {
        var redisId, _socket;
        switch (data.type) {
          case 'MOVED':
            redisId = data.message.replace(/.+\s/, '');
            redisState.slots[command.slot] = redisId;
            _socket = addSocket(redis, redisId);
            _socket.commandQueue.push(command);
            execQueue(redisState, _socket);
            break;

          case 'ASK':
            redisId = data.message.replace(/.+\s/, '');
            _socket = addSocket(redis, redisId);
            _socket.commandQueue.push(createCommand(redis, ['asking'], function(error, res) {
              if (error) return command.callback(error);
              _socket.commandQueue.push(command);
              execQueue(redisState, _socket);
            }));
            execQueue(redisState, _socket);
            break;

          default:
            command.callback(data);
        }

        return redis.emit('warn', data);
      }

      if (command.name === 'monitor') {
        redisState.monitorMode = true;
        this.setAutoEnd();
        return command.callback(null, data);
      }

      if (isSubReply(data)) {
        // (pub)subscribe can generate many replies. All are emitted as events.
        if (!redisState.pubSubMode) {
          redisState.pubSubMode = true;
          this.setAutoEnd();
        }
        command.callback();
        return redis.emit.apply(redis, data);
      }

      return command.callback(null, data);
    })
    .on('end', function() {
      socket.pendingWatcher = null;
      execQueue(redisState, socket);
    });
}

// This Command constructor is ever so slightly faster than using an object literal, but more importantly, using
// a named constructor helps it show up meaningfully in the V8 CPU profiler and in heap snapshots.
function Command(command, slot, data, callback, additionalCallbacks) {
  this.slot = slot;
  this.data = data;
  this.name = command;
  this.callback = callback;
  this.additionalCallbacks = additionalCallbacks || 0;
}

function createCommand(redis, reqArray, callback, additionalCallbacks) {
  var buffer, slot;
  try {
    slot = redis.clientCalcSlot(reqArray);
    buffer = resp.bufferify(reqArray);
  } catch (error) {
    return callback(error);
  }
  return new Command(reqArray[0], slot, buffer, callback, additionalCallbacks);
}

function execCommands(socket, commandQueue, pendingWatcher) {
  socket.pendingWatcher = pendingWatcher || {commands: new Queue()};
  pendingWatcher = socket.pendingWatcher;
  var count = socket.commandsHighWater;

  while (commandQueue.length && count--) {
    var command = commandQueue.shift();
    if (socket.debugMode) tool.log({socketWrite: command.data.toString()});

    pendingWatcher.commands.push(command);

    while (command.additionalCallbacks--)
      pendingWatcher.commands.push({
        name: command.name,
        callback: noOp
      });

    if (!socket.write(command.data)) break;
  }
}

function sendCommand(redis, command, args, additionalCallbacks) {
  return Thunk.call(redis, function(callback) {
    if (this._redisState.ended) return callback(new Error('The redis client was ended'));
    args = tool.slice(args);
    args.unshift(command);
    dispatchCommands(this, createCommand(redis, args, callback, additionalCallbacks));
  });
}

function dispatchCommands(redis, command) {
  var redisState = redis._redisState;
  var commandQueue = redisState.commandQueue;

  if (!redisState.connected) {
    if (command) commandQueue.push(command);
    return;
  }

  var count = commandQueue.length;

  if (!command && !count) return;
  if (command && !count) return execQueue(redisState, dispatchCommand(redisState, command).socket);

  var assignedSockets = Object.create(null);
  var matched = null;
  while (commandQueue.length) {
    matched = dispatchCommand(redisState, commandQueue.shift());
    assignedSockets[matched.redisId] = matched.socket;
  }

  if (command) {
    matched = dispatchCommand(redisState, command);
    assignedSockets[matched.redisId] = matched.socket;
  }
  tool.each(assignedSockets, function(socket) {
    execQueue(redisState, socket);
  });
}

function dispatchCommand(redisState, command) {
  var redisId = redisState.slots[command.slot];
  var socket = redisState.pool[redisId];
  if (!socket) throw new Error(redisId + ' is not connected');
  socket.commandQueue.push(command);
  return {socket: socket, redisId: socket.redisId};
}

function execQueue(redisState, socket) {
  if (!socket.connected || !socket.commandQueue.length) return;
  var continuous = redisState.pubSubMode || redisState.monitorMode;
  if (!continuous && socket.pendingWatcher) return;
  execCommands(socket, socket.commandQueue, continuous ? socket.pendingWatcher : {commands: new Queue()});
}

function flushPending(socket, error) {
  if (!socket.pendingWatcher) return;
  var command = socket.pendingWatcher.commands.shift();
  while (command) {
    command.callback(error);
    command = socket.pendingWatcher.commands.shift();
  }
  socket.pendingWatcher = null;
}

function flushToRedis(redis, socket, error) {
  flushPending(socket, error);
  var redisQueue = redis._redisState.commandQueue;
  var command = socket.commandQueue.shift();
  while (command) {
    redisQueue.push(command);
    command = socket.commandQueue.shift();
  }
}

function unmountSocket(redis, socket, hadError, tryEnd) {
  if (!socket.connected) return;
  var redisState = redis._redisState;
  socket.connected = false;
  socket.destroy();
  delete redisState.pool[socket.redisId];
  redis.emit('warn', new Error(hadError || (socket.redisId + ' disconnected')));
  var sockets = Object.keys(redisState.pool);
  if (!sockets.length && tryEnd) return redis.clientEnd(hadError);
  if (redisState.slots[-1] !== socket.redisId) return;
  // reset default socket
  redisState.slots[-1] = null;
  for (var i = 1; i < sockets.length; i++) {
    if (redisState.pool[sockets[i]].redisMaster) {
      redisState.slots[-1] = sockets[i];
      return;
    }
  }
  redisState.slots[-1] = sockets[0];
}

var messageTypes = Object.create(null);
messageTypes.message = true;
messageTypes.pmessage = true;

function isMessageReply(reply) {
  return reply && messageTypes[reply[0]];
}

var subReplyTypes = Object.create(null);
subReplyTypes.subscribe = true;
subReplyTypes.psubscribe = true;

function isSubReply(reply) {
  return reply && subReplyTypes[reply[0]];
}

var unSubReplyTypes = Object.create(null);
unSubReplyTypes.unsubscribe = true;
unSubReplyTypes.punsubscribe = true;

function isUnSubReply(reply) {
  return reply && unSubReplyTypes[reply[0]];
}

function noOp() {}
